docs/book/tutorials/kubernetes-on-vsphere-with-kubeadm.md

```diff
@@ -156,262 +208,994 @@ bootstrapTokens:
+kube-flannel-ds-amd64-j7g4s              1/1     Running   0          22m
+kube-flannel-ds-amd64-q4zsn              1/1     Running   0          21m
+kube-proxy-6jcng                         1/1     Running   0          30m
+kube-proxy-bh8kh                         1/1     Running   0          21m
+kube-proxy-rb9xp                         1/1     Running   0          22m
+kube-proxy-srhpj                         1/1     Running   0          71m
+kube-proxy-vh4lg                         1/1     Running   0          38m
+kube-scheduler-k8s-master                1/1     Running   0          70m
+vsphere-cloud-controller-manager-549hb   1/1     Running   0          25s
+```
+Check that all nodes are untainted
+Verify node.cloudprovider.kubernetes.io/uninitialized taint is removed from all nodes.
+```
+# kubectl describe nodes | egrep "Taints:|Name:"
+Name:               k8s-master
+Taints:             node-role.kubernetes.io/master:NoSchedule
+Name:               k8s-node1
+Taints:             <none>
+Name:               k8s-node2
+Taints:             <none>
+Name:               k8s-node3
+Taints:             <none>
+Name:               k8s-node4
+Taints:             <none>
+```
+
+Note: If you happen to make an error with the vsphere.conf, simply delete the CPI components and the configMap, make any necessary edits to the configMap vSphere.conf file, and reapply the steps above.
 
-### Verify setup
+## Install vSphere Container Storage Interface Driver
+Now that the CPI is installed, we can focus on the CSI. Perform the following steps on the Master Node(s) only.
+Taint the master nodes to prevent scheduling
+The node-role.kubernetes.io/master=:NoSchedule taint is required to be present on the master nodes to prevent scheduling of the node plugin pods for vsphere-csi-node daemonset on the master nodes.
+```
+# kubectl taint nodes k8s-master node-role.kubernetes.io/master=:NoSchedule
+```
+### Create a CSI secret
+Just like we saw with the CPI, this credential secret, passed to the CSI on initialization, contains details about the vSphere configuration. This csi-vsphere.conf has been populated with some sample values. Obviously, you will need to modify this file to reflect your own vSphere configuration.
 
-Now, as the output says above, back on the master check that all nodes have joined the cluster
+For information about using a certificate and private key instead of a username and password in plain text, see Use Certificate to Secure vCenter Server Connections.
+```
+# tee /etc/kubernetes/csi-vsphere.conf >/dev/null <<EOF
+[Global]
+cluster-id = "demo-cluster-id"  
+[VirtualCenter "1.1.1.1"]
+insecure-flag = "true"
+user = "Administrator@vsphere.local"
+password = "password"
+port = "443"
+datacenters = "datacenter"
+EOF
+```
+The entries in the csi-vsphere.conf file have the following meanings:
+cluster-id represents the unique cluster identifier. Each kubernetes cluster should have it's own unique cluster-id set in the csi-vsphere.conf file.
```

Comment:
Highlight this is very important to avoid PV name collisions.

Reply:
Done